pkg/tinydtls/contrib/sock_dtls.c

```diff
@@ -376,29 +442,81 @@ static ssize_t _copy_buffer(sock_dtls_t *sock, void *data, size_t max_len)
     if (buflen > max_len) {
         return -ENOBUFS;
     }
+#if SOCK_HAS_ASYNC
+    if (sock->buf_ctx != NULL) {
+        memcpy(data, buf, sock->buflen);
+        _check_more_chunks(sock->udp_sock, (void **)&buf, &sock->buf_ctx,
+                           &remote->ep);
+        if (sock->async_cb &&
+            /* is there a message in the sock's mbox? */
+            cib_avail(&sock->mbox.cib)) {
+            if (sock->buf) {
+                sock->async_cb(sock, SOCK_ASYNC_MSG_RECV,
+                               sock->async_cb_arg);
+            }
+            else {
+                sock->async_cb(sock, SOCK_ASYNC_CONN_RECV,
+                               sock->async_cb_arg);
+            }
+        }
+        return buflen;
+    }
+#else
+    (void)remote;
+#endif
     /* use `memmove()` as tinydtls reuses `data` to store decrypted data with an
      * offset in `buf`. This prevents problems with overlapping buffers. */
     memmove(data, buf, buflen);
     return buflen;
 }
 
+static ssize_t _complete_handshake(sock_dtls_t *sock,
+                                   sock_dtls_session_t *remote,
+                                   const session_t *session)
+{
+    memcpy(&remote->dtls_session, session, sizeof(remote->dtls_session));
+#ifdef SOCK_HAS_ASYNC
+    if (sock->async_cb) {
+        sock_async_flags_t flags = SOCK_ASYNC_CONN_RDY;
+
+        if (cib_avail(&sock->mbox.cib)) {
+            if (sock->buf) {
+                flags |= SOCK_ASYNC_MSG_RECV;
+            }
+            else {
+                flags |= SOCK_ASYNC_CONN_RECV;
+            }
+        }
+        sock->async_cb(sock, flags, sock->async_cb_arg);
+    }
+#else
+    (void)sock;
+#endif
+    return -SOCK_DTLS_HANDSHAKE;
+}
+
 ssize_t sock_dtls_recv(sock_dtls_t *sock, sock_dtls_session_t *remote,
                        void *data, size_t max_len, uint32_t timeout)
 {
     assert(sock);
     assert(data);
     assert(remote);
 
-    if (sock->buf != NULL) {
-        /* there is already decrypted data available */
-        return _copy_buffer(sock, data, max_len);
-    }
-
     /* loop breaks when timeout or application data read */
     while (1) {
+        ssize_t res;
         uint32_t start_recv = xtimer_now_usec();
-        ssize_t res = sock_udp_recv(sock->udp_sock, data, max_len, timeout,
-                                    &remote->ep);
+        msg_t msg;
+
+        if (sock->buf != NULL) {
+            return _copy_buffer(sock, remote, data, max_len);
+        }
+        else if (mbox_try_get(&sock->mbox, &msg) &&
+                 msg.type == DTLS_EVENT_CONNECTED) {
+            return _complete_handshake(sock, remote, msg.content.ptr);
+        }
```

Comment:
also see https://github.com/RIOT-OS/RIOT/pull/12907#discussion_r441399367.

I'm suspect the condition we are checking here is the issue, I think for the second message, `sock->buf` can be not `NULL` while there is already a `DTLS_EVENT_CONNECTED`, meaning the message is already decrypted and saved to `sock->buf` while the user haven't yet called `sock_dtls_recv()` to accept the connection (return `SOCK_DTLS_HANDSHAKE`). I assume the decryption is done by the `dtls_handle_message()` call in `_udp_cb`.

Reply:
Mhhh maybe we should have a `mbox_peek()` that looks up the message, but does not pop it from the queue, so we can be sure what is on top of the queue and act accordingly, instead of checking side effects.